infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java

```diff
@@ -219,11 +225,16 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
     /**
      * Resend the messages in the queue if they have timed out.
      */
-    public LogReplicationEntryMsg resend() {
+    public LogReplicationEntryMsg resend(boolean isLogEntry) {
         LogReplicationEntryMsg ack = null;
         boolean force = false;
+
         try {
             ack = processAcks();
+
+            if (isLogEntry && isBackpressureActive.get()) {
+                waitBeforeRetry();
+            }
```

Comment:
I think a better back pressure mechanism would do the following:
1. consumer rejects message explicitly, returns busy response (as opposed to waiting for a timeout)
2. producer waits, then sends a probe message without the payload
3. if the probe returns not-busy, then the producer sends the payload, else keep probing

Reply:
We had considered something similar by bounding the executor service that holds the RPCs on the sink, where any message beyond capacity is met will a null ack. I agree that having explicit messaging is a better approach but one objective of this fix was to minimally modify the current behavior, the introduction of additional RPCs might require more care.

Reply:
yes, adding a 'busy' response on the receiver and probe request from the sender will require a more involved change which we had decided to avoid.